java-frontend/src/main/java/org/sonar/java/se/checks/NonNullSetToNullCheck.java

```diff
@@ -0,0 +1,237 @@
+      }
+    }
+
+    @Override
+    public void visitMethodInvocation(MethodInvocationTree syntaxTree) {
+      Symbol symbol = syntaxTree.symbol();
+      if (symbol.isMethodSymbol()) {
+        Arguments arguments = syntaxTree.arguments();
+        List<SymbolicValue> argumentValues = new ArrayList<>(programState.peekValues(arguments.size() + 1));
+        argumentValues.remove(0);
+        Collections.reverse(argumentValues);
+        JavaSymbol.MethodJavaSymbol methodSymbol = (JavaSymbol.MethodJavaSymbol) symbol;
+        checkNullArguments(syntaxTree, methodSymbol.getParameters(),
+          argumentValues, "Parameter {0} to this call is marked \"{1}\" but null is passed.");
+      }
+    }
+
+    protected void checkNullArguments(Tree syntaxTree, Scope parameters, List<SymbolicValue> argumentValues, String message) {
+      if (parameters != null) {
+        List<JavaSymbol> scopeSymbols = parameters.scopeSymbols();
+        int parametersToTest = argumentValues.size();
+        if (scopeSymbols.size() < parametersToTest) {
+          // The last parameter is a variable length argument: the non-null condition does not apply to its values
+          parametersToTest = scopeSymbols.size() - 1;
+        }
+        for (int i = 0; i < parametersToTest; i++) {
+          checkNullArgument(syntaxTree, message, scopeSymbols, argumentValues, i);
+        }
+      }
+    }
+
+    protected void checkNullArgument(Tree syntaxTree, String message, List<JavaSymbol> scopeSymbols, List<SymbolicValue> argumentValues, int i) {
+      String nonNullAnnotation = nonNullAnnotation(scopeSymbols.get(i));
+      if (nonNullAnnotation != null) {
+        Constraint constraint = programState.getConstraint(argumentValues.get(i));
+        if (constraint != null && constraint.isNull()) {
+          reportIssue(syntaxTree, message, Integer.valueOf(i + 1), nonNullAnnotation);
+        }
+      }
+    }
+  }
+
+  private class PostStatementVisitor extends AbstractStatementVisitor {
+
+    protected PostStatementVisitor(CheckerContext context) {
+      super(context);
+    }
+
+    @Override
+    public void visitReturnStatement(ReturnStatementTree tree) {
+      Tree parent = tree.parent();
+      while (!parent.is(Tree.Kind.METHOD)) {
+        parent = parent.parent();
+        if (parent == null) {
+          // This occurs when the return statement is within a constructor
+          return;
+        }
+      }
+      MethodTree methodTree = (MethodTree) parent;
+      String nonNullAnnotation = nonNullAnnotation((JavaSymbol) methodTree.symbol());
+      if (nonNullAnnotation != null) {
+        SymbolicValue returnedValue = programState.peekValue();
+        Constraint constraint = programState.getConstraint(returnedValue);
+        if (constraint == null) {
+          reportIssue(tree, "This method''s return value is marked \"{0}\" but null may be returned.", nonNullAnnotation);
```

Comment:
I'm not sure about reporting that issue. What about that case; The field is final and set in the constructor:

```
class MyOtherClass {
  private final Object myObject;
  
  public MyOtherClass() {
    myObject = new Object();
  }
  
  @Nonnull
  public Object getMyObject() {
    return myObject; // would raise a FP
  }
}
```